devices/nand-imx6ull/raw.c

```diff
@@ -0,0 +1,159 @@
+
+#include "drv.h"
+#include "data.h"
+
+/* NOTICE: raw device relies on data device for init and done methods. */
+
+
+/* Linker symbols */
+extern u8 nand_raw_page[]; /* NAND page cache */
+
+
+static ssize_t raw_read(unsigned int minor, addr_t offs, void *buff, size_t len, time_t timeout)
+{
+	nand_t *nand = nand_get(minor);
+	size_t size, rawEraseBlockSz, rawsz, rawPagesz, ret, page, poffs;
+	int err;
+
+	if (nand == NULL) {
+		return -ENODEV;
+	}
+
+	rawEraseBlockSz = (nand->cfg->erasesz / nand->cfg->writesz) * (nand->cfg->writesz + nand->cfg->metasz);
+	rawsz = (nand->cfg->size / nand->cfg->erasesz) * rawEraseBlockSz;
+	rawPagesz = nand->cfg->writesz + nand->cfg->metasz;
+
+	if (((len != 0u) && (buff == NULL)) || (offs >= rawsz)) {
+		return -EINVAL;
+	}
+
+	len = min(len, rawsz - offs);
+	if (len == 0u) {
+		return 0;
+	}
+
+	/* Sync data device changes. */
+	err = data_doSync(nand);
+	if (err < 0) {
+		return err;
+	}
+
+	for (ret = 0; ret < len; ret += size) {
+		page = (offs + ret) / rawPagesz;
+		poffs = (offs + ret) % rawPagesz;
+		size = min(len - ret, rawPagesz - poffs);
+
+		err = flashdrv_readraw(nand->dma, page, nand_raw_page, rawPagesz);
+		if (err < 0) {
+			return err;
+		}
+		hal_memcpy((u8 *)buff + ret, nand_raw_page + poffs, size);
+	}
+
+	return ret;
+}
+
+
+static ssize_t raw_write(unsigned int minor, addr_t offs, const void *buff, size_t len)
+{
+	nand_t *nand = nand_get(minor);
+	size_t size, rawEraseBlockSz, rawsz, rawPagesz, ret;
+	int err;
+
+	if (nand == NULL) {
+		return -ENODEV;
+	}
+
+	rawEraseBlockSz = (nand->cfg->erasesz / nand->cfg->writesz) * (nand->cfg->writesz + nand->cfg->metasz);
+	rawsz = (nand->cfg->size / nand->cfg->erasesz) * rawEraseBlockSz;
+	rawPagesz = nand->cfg->writesz + nand->cfg->metasz;
+
+	if (((len != 0u) && (buff == NULL)) || (offs >= rawsz) ||
+		((rawsz % rawPagesz) != 0u) || ((offs % rawPagesz) != 0u)) {
+		return -EINVAL;
+	}
+
+	len = min(len, rawsz - offs);
+	if (len == 0u) {
+		return 0;
+	}
+
+	/* Sync data device changes. */
+	err = data_doSync(nand);
+	if (err < 0) {
+		return err;
+	}
+
+	for (ret = 0; ret < len; ret += size) {
+		size = min(len - ret, rawPagesz);
+
+		hal_memcpy(nand_raw_page, (u8 *)buff + ret, size);
+		err = flashdrv_writeraw(nand->dma, (offs + ret) / rawPagesz, nand_raw_page, size);
+		if (err < 0) {
+			return err;
+		}
+	}
+
+	return ret;
+}
+
+
+static int raw_sync(unsigned int minor)
+{
+	/* Raw device has no buffers to sync. */
+	return 0;
+}
+
+
+static int raw_done(unsigned int minor)
+{
+	/* Device deinit is done in data_done */
+	return 0;
+}
+
+
+static int raw_init(unsigned int minor)
+{
+	nand_t *nand = nand_get(minor);
+
+	if (nand == NULL) {
+		return -EPERM;
+	}
+	lib_printf("\ndev/nand/raw: Configured %s(%d.%d)", nand->cfg->name, DEV_NAND_RAW, minor);
+
+	return 0;
+}
+
+
+__attribute__((constructor)) static void raw_register(void)
+{
+	static const dev_handler_t h = {
+		.read = raw_read,
+		.write = raw_write,
+		.sync = raw_sync,
```

Comment:
sync is somewhat optional, i.e. there's a NULL check: https://github.com/phoenix-rtos/plo/blob/642ac8503a6de0bc2ac9f46a95bc188a9edc8540/devices/devs.c#L140

I guess it is problematic for it to return `-EINVAL` on use?

Reply:
Applies to sync and done for both meta and raw ifaces